### ISSUE TITLE
OKTA IDP doesn't work for version 0.10.2

### ISSUE THREAD

**Issue by KevDBG** [pomerium/pomerium#1341]
Hello,

## What happened?

We follow this documentation to deploy pomerium on a K8S cluster (rancher RKE) -> https://github.com/pomerium/pomerium/tree/v0.10.2/examples/kubernetes.

Our configuration works with version 0.8.x and we tried to upgrade to the last version v0.10.2. The OKTA authentication works, but the access to the backend application defined in the policy is denied.

We saw a change between these version on the `idp_service_account `, now it's seems that we have to use a json base64 encode. So we did that:

```
cat api_key.json

{
    "api_token": "XXXX OKTA IDP TOKEN XXXX"
}

cat api_key.json | base64
ewogICAgImFwaV90b2tlbiI6ICIiWFhYWCBPS1RBIElEUCBUT0tFTiBYWFhYCn0K

```
We use this in the pomerium config.yml file (we also try with "api_key" format with no success, we find this kind of key in a commit).

The authentication works but we have a denied message, we are pretty sure that the call API to OKTA doesn't work but we don't find a log linked to this behavior (excepted the log below) on all pomerium services (we use separated services).

![image](https://user-images.githubusercontent.com/6896709/91484991-5c5e7500-e8aa-11ea-8974-e1b37a3b296d.png)

## What did you expect to happen?

Pomerium regarding the defined policy should accept the connection (we have the same configuration on each component in 0.8.X and it's works).

## What's your environment like?

- Pomerium version -> 0.10.2
- Server Operating System/Architecture/Cloud: Centos8 / K8S by RKE rancher

## What's your config.yaml?

Our kubernetes-config.yaml is:
```
insecure_server: true
grpc_insecure: true
grpc_address: ":80"

pomerium_debug: true
authenticate_service_url: https://authenticate.sso.domain.tld
authorize_service_url: http://pomerium-authorize-service.namespace.svc.cluster.local
cache_service_url: http://pomerium-cache-service.namespace.svc.cluster.local

override_certificate_name: "*.sso.domain.tld"

idp_provider: okta
idp_client_id: <OKTA_APP_CLIENT_ID>
idp_client_secret: <OKTA_APP_CLIENT_SECRET>
idp_provider_url: https://ourdomain.okta.com
idp_service_account: ewogICAgImFwaV90b2tlbiI6ICIiWFhYWCBPS1RBIElEUCBUT0tFTiBYWFhYCn0K

policy:
    - from: https://hello.sso.domain.tld
      to: http://hello.namespace.svc.cluster.local
      allowed_groups:
        - <OKTA_GROUP_ID>
```

## What did you see in the logs?

I don't know if it's link but we have this kind of log during the authentication:

```
7:14PM INF authenticate: session load error error="Bad Request: internal/sessions: session is not found" X-Forwarded-For=["10.42.32.0,10.42.32.11"] X-Forwarded-Host=["authenticate.sso.domain.tld"] X-Forwarded-Port=["443"] X-Forwarded-Proto=["http"] X-Forwarded-Server=["traefik-ingress-controller-external-68c79c4f8-p8w4x"] X-Real-Ip=["10.42.32.0"] ip=127.0.0.1 request-id=XXXXX user_agent="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"
7:14PM INF authenticate: session load error error="Bad Request: internal/sessions: session is not found" X-Forwarded-For=["10.42.32.0,10.42.32.11"] X-Forwarded-Host=["authenticate.sso.domain.tld"] X-Forwarded-Port=["443"] X-Forwarded-Proto=["http"] X-Forwarded-Server=["traefik-ingress-controller-external-68c79c4f8-p8w4x"] X-Real-Ip=["10.42.32.0"] ip=127.0.0.1 request-id=YYYYY user_agent="Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36"
```
Thanks for your help,

**Comment by desimone** [comment]
@KevDBG 

Thanks for the detailed issue. Could you provide a bit more logging context? In particular, I'm looking to see what type of logs are coming out of the cache service to see if groups are being pulled from Okta successfully.

If you go to the request details on that authorization denied page, what do you see? Are the user's groups listed?

Thanks!
Bobby

**Comment by KevDBG** [comment]
Hey @desimone !

Thanks for this quick take in charge.

You have right ! I didn't check the cache service because i beleaved that is disappear in future release, so it was not useful (it's a mistake ^^). Indeed, there are interesting logs about OKTA :+1 in the service cache log: 

```
7:39PM INF getting user groups provider=okta service=directory
7:40PM WRN failed to refresh directory users and groups error="okta: error querying for groups: rate: Wait(n=1) would exceed context deadline" service=identity_manager
7:40PM INF refreshing user service=identity_manager user_id=okta/XXXXXXXXXXXX
7:40PM INF set id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:40PM INF set id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:50PM INF refreshing directory users service=identity_manager
7:50PM INF getting user groups provider=okta service=directory
7:51PM WRN failed to refresh directory users and groups error="okta: error querying for groups: rate: Wait(n=1) would exceed context deadline" service=identity_manager
7:51PM INF refreshing user service=identity_manager user_id=okta/XXXXXXXXXXXX
7:51PM INF set id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:51PM INF set id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:55PM INF get id=ID.inhp-ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:55PM INF set id=ID.c-ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:55PM INF set id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=ID.ZZZZZZZZZZZZZZZZZZZZ service=databroker type=type.googleapis.com/session.Session
7:55PM INF get id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/user.User
7:55PM INF get id=okta/XXXXXXXXXXXX service=databroker type=type.googleapis.com/directory.User
```
(my login test is linked to log at 7:55PM)
XXXXXXXXXXXX is my OKTA user ID.

When i checked the page https://authenticate.sso.domain.tld/.pomerium/? i have:

```
Name
Firstname lastname
UserID
okta/XXXXXXXX
Email
foo@xxx.com
Expires At
2020-08-28 09:55:12.977331027 +0000 UTC
Issued At
2020-08-27 19:55:12 +0000 UTC
Sign Out
```

So, nope the groups isn't listed. Any idea with these new elements ?

**Comment by KevDBG** [comment]
@desimone Hey !

It's seems that we have a problem around the go limiter (rate limiting)... I don't understand why, we are sure that our token is correct.

I just try with api_key in json regarding our version 0.10.2 and https://github.com/pomerium/pomerium/blob/v0.10.2/internal/directory/okta/okta.go#L283

**Comment by KevDBG** [comment]
We have almost one hundred okta group, so with the loop on each group we reach maybe some rate limiting in your code :

`const defaultQPS = 100 / 60`

We have an okta enterprise account, so maybe you can add an environment variable to customize this value ?

**Comment by KevDBG** [comment]
New tests:

We tried with 0.10.1 and **it's works**. So, we have a problem with this https://github.com/pomerium/pomerium/pull/1271

In 0.10.2 we tried to fix idp_qps to 60.0 with different way (config.yaml or environment variable) and it's doesn't work for us.

It's cool because we can move forward on 0.10.1. We can provide you more test if you need to troubleshoot in 0.10.2.

**Comment by cuonglm** [comment]
@KevDBG The thing is that the ability to config QPS is not backported, so you can't config it.

@travisgroth is there any problem that we can't backport the change to config QPS?

**Comment by travisgroth** [comment]
@KevDBG not ideal, but I believe you could also extend the timeout setting `idp_refresh_directory_timeout` from the default of 1m.

@cuonglm I don't think there is a technical problem, but it was left out since introducing new options isn't something we typically do in patch releases.  Can discuss further if we need an exception (or better default) here for a v0.10.3.

**Comment by KevDBG** [comment]
Hello,

@cuonglm , i don't understand your return `@KevDBG The thing is that the ability to config QPS is not backported, so you can't config it.`.

Indeed, we tested the version 0.10.2 which include the idp_qps parameter (regarding this https://github.com/pomerium/pomerium/pull/1271) and it's doesn't work (we reach the default limit and the paramater is not taken into account). So, maybe i'm wrong but i don't understand the backport mention.

You think maybe, that we had tested the parameter idp_qps in version 0.10.1.

Thanks for your help,

**Comment by cuonglm** [comment]
@KevDBG I mean the code that allowing set `idp_qps` is not backported, so you can't change the default value of `1`.

**Comment by cuonglm** [comment]
@KevDBG Would you mind increasing the refresh timeout/interval to see if the problem still occurs?

https://www.pomerium.io/reference/#identity-provider-refresh-directory-settings

**Comment by devstein** [comment]
I am seeing a similar, possibly the same issue, in `0.10.2` with Okta where Pomerium is getting empty `groups` for all new users. It _seems_ like it could be caching related. Is this a separate issue?

**Comment by desimone** [comment]
@devstein / @KevDBG  -- Are you able to test this on v0.10.6?

**Comment by bazzuka** [comment]
I am seeing still empty groups in 0.10.6.

**Comment by desimone** [comment]
@bazzuka -- could you provide your redacted configuration?

**Comment by bazzuka** [comment]
```
# For detailed explanation of each of the configuration settings see
# https://www.pomerium.io/reference/

nameOverride: ""
fullnameOverride: ""

# settings that are shared by all services
config:
  # routes under this wildcard domain are handled by pomerium
  rootDomain: "domain.com"
  existingSecret: ""
  existingCASecret: ""
  ca:
    cert: ""
    key: ""
  sharedSecret: ""
  cookieSecret: ""
  generateTLS: false
  generateTLSAnnotations: {}
  forceGenerateTLS: false
  generateSigningKey: true
  forceGenerateSigningKey: false
  extraOpts: {}
  existingPolicy: ""
  insecure: true
  administrators: ""
  policy: 
    - from: https://domain.1
      to: http://lenses.lenses.svc.cluster.local
      allowed_users:
        - user@domain.com
      allowed_groups:
        - devops
  existingSigningKeySecret: ""
  signingKey: ""

authenticate:
  fullnameOverride: ""
  nameOverride: ""
  existingTLSSecret: ""
  redirectUrl: ""
  cacheServiceUrl: ""
  # see https://www.pomerium.io/docs/identity-providers.html
  idp:
    provider: okta
    clientID: "client_id"
    clientSecret: "client_secret"
    url: "https://company.okta.com"
    refresh_directory_interval: 5m
    serviceAccount: "apitoken"
  tls:
    cert: ""
    key: ""
    defaultSANList: []
    defaultIPList: []
  replicaCount: 3
  autoscaling:
    enabled: false
    minReplicas: 1
    maxReplicas: 5
    targetCPUUtilizationPercentage: 50
    targetMemoryUtilizationPercentage: 50
```

**Comment by bazzuka** [comment]
<img width="822" alt="Screen Shot 2020-11-08 at 3 07 15 PM" src="https://user-images.githubusercontent.com/15695060/98466070-e5145e00-21d5-11eb-9b1b-bcccdab1e384.png">
<img width="996" alt="Screen Shot 2020-11-08 at 3 17 47 PM" src="https://user-images.githubusercontent.com/15695060/98466090-f8272e00-21d5-11eb-8be0-9bd451d25295.png">
<img width="756" alt="Screen Shot 2020-11-08 at 3 18 20 PM" src="https://user-images.githubusercontent.com/15695060/98466096-02e1c300-21d6-11eb-9dbc-26e008d40eee.png">
<img width="817" alt="Screen Shot 2020-11-08 at 3 21 24 PM" src="https://user-images.githubusercontent.com/15695060/98466112-1ab94700-21d6-11eb-9b48-1d9c69cf13bc.png">

**Comment by bazzuka** [comment]
```
{"level":"info","service":"authorize","request-id":"dc23e322-9a74-4143-81fa-5d7a497d6ed0","check-request-id":"b80c3766-49f1-44e1-acb8-f784f7425698","method":"GET","path":"/login.html?redirect=/","host":"lenses2.domain.com","query":"","allow":false,"status":403,"message":"forbidden","user":"vladyslavmakalish@domain.com","groups":[],"time":"2020-11-08T13:17:32Z","message":"authorize check"}
```

**Comment by desimone** [comment]
@bazzuka do you mind posting a complete (but redacted) gist of your logs?

**Comment by bazzuka** [comment]
[logs.txt](https://github.com/pomerium/pomerium/files/5517498/logs.txt)
@desimone

**Comment by calebdoxsey** [comment]
@bazzuka is your `idp_service_account` setup with an API token in this format:

```json
{
  "api_token": "..."
}

```

https://www.pomerium.io/docs/identity-providers/okta.html#service-account ?

If the databroker is unable to sync groups there should be an error message on startup. It'd be in the logs of the databroker (cache) service. Do you see anything there?

**Comment by bazzuka** [comment]
idp:
 serviceAccount: (output of base64 -i token.json with proper format)
```
{"level":"warn","service":"directory","provider":"okta","error":"okta: api_key is required","time":"2020-11
-29T14:41:55Z","message":"invalid service account for okta directory provider"}
{"level":"warn","provider":"okta","time":"2020-11-29T14:41:55Z","message":"no directory provider implementa
tion found, disabling support for groups"}
```

**Comment by bazzuka** [comment]
idp:
 serviceAccount: (output of base64 -i token.json with proper format)
```
{"level":"warn","service":"directory","provider":"okta","error":"okta: api_key is required","time":"2020-11
-29T14:41:55Z","message":"invalid service account for okta directory provider"}
{"level":"warn","provider":"okta","time":"2020-11-29T14:41:55Z","message":"no directory provider implementa
tion found, disabling support for groups"}
```

**Comment by bazzuka** [comment]
fixed by adding instead of api_token
{
    "api_key": "my_key"
}
```
type ServiceAccount struct {
339		APIKey string `json:"api_key"`
340	}
```

**Comment by bazzuka** [comment]
@desimone I think we can close this issue, just update json key in documentation.

**Comment by calebdoxsey** [comment]
@bazzuka it appears updated on the master branch so should get fixed when we release 0.11.
